gnucash/gnome/dialog-trans-assoc.c

```diff
@@ -144,7 +144,7 @@ convert_uri_relative_to_uri (AssocDialog *assoc_dialog, const gchar *uri)
 {
     const gchar *new_uri;
 
-    if (assoc_dialog->valid_path_head && g_str_has_prefix (uri,"file:/") &&
+    if (assoc_dialog->valid_path_head && g_str_has_prefix (uri,"file:") &&
```

Comment:
You are correct https://tools.ietf.org/html/rfc8089#section-3 doesn't specify relative file uri's. However our code needs a way to handle those anyway.

Unfortunately your code has chosen a different convention from mine in gnc-uri-utils.c. This is a potential source of confusion and errors. That's what I'm pointing out. I won't claim my work is right. I have written the uri code a long time ago and most likely was pretty naive while writing it.

For starters it won't parse a localhost intermediate. It will simply become part of the local path name. This is really a bug, but I don't know if it matters in practice.

Your choice for relative file uri's is `file:/my-shared-text-file.txt` while mine is `file://my-shared-text-file.txt`. Clearly my scheme will lead to ambiguity if the relative path would be `file://hostname/my-shared-text-file.txt` so retrospect this is an unfortunate choice. If I remember correctly I made that choice to universally be able to strip `file://` from the URI to get either an absolute or a relative path.

For consistency I think we should abandon at least my choice of a relative scheme for file handling. If a file URI is passed to GnuCash, GnuCash should resolve it as if it's absolute. If that results in a "file not found" this should be reported to the user as such. On the other hand, GnuCash internally keeps track of the book's location in the form of a file URI. So if the path passed to GnuCash was relative, GnuCash can still make it absolute using some conversion routine before making it a URI. That way our internal URI is always consistent.

Similarly while looking at your use case I doubt we need to support something like `file:/somefile.txt`. It would be better to also unambiguously store either a proper file uri (meaning with absolute path) or simply a relative path without uri scheme. Your code can easily detect whether there's a uri scheme in the stored string and act accordingly. No scheme: assume this is a local path name and determine absoluteness before adding a potential path head. With file scheme, consider path always absolute. You are using the GLib uri functions to parse file schemes, so it should do the right thing for you.

Perhaps this is also what you meant with 
> (that probably should change which I was contemplating along with some other changes later)

The unfortunate part is that we currently have code in the wild that uses these contrived schemes. I haven't studies your code in much more detail, so I wonder in which circumstances will it store a relative path using the `file:/path/to/file.txt` scheme and when `path/to/file.txt` ?

Reply:
At the moment, the stored value for absolute file paths is 'file:///usr/share/file.txt' or for a relative path either 'file:/my-file.txt' or 'file:my-file.txt' depending on what the path head was set to.

If the 'transaction location association' was changed to only accept uri's, then any thing else can be treated as a relative path so the following should work when the association is retrieved...
Test for 'file:' but not 'file://', remove the 'file:' part and save this stripped value.
Test for a scheme with g_uri_parse_scheme,

- if a scheme but not 'file' then treat as location uri
- If a file scheme then it is absolute
- If it is not a scheme, add it to what ever the path head is for an absolute path.

Over time all entries with the old format would get changed and if they used the 'Transaction Association' dialogue they would all get changed as part of the retrieval process.

Reply:
This is a bit skew to the discussion at hand, but I'd prefer that you use/extend the URI handling in libgnucash/engine/gnc-uri-utils.h to using GLib ones. I think those are the ones that @gjanssens is referring to. Doing so will promote better uniformity across GnuCash and make it easier to migrate away from GLib. For the code that @gjanssens originally commented on that would be ```gnc_uri_is_file_uri```; instead of ```g_uri_parse``` use ```gnc_uri_get_components```.